### ISSUE TITLE
(fuera de tema) duda con vector dinámico

### ISSUE THREAD

**Issue by teofum** [PI-ITBA/2024_01#133]
La consulta se va de los temas que estamos viendo en clase, pero estaba viendo las clases y guía de structs y tuve la idea de armar un struct `vector` genérico, que permita crear un vector dinámico de elementos de cualquier tipo. Para eso hice este struct y funciones:

```c
#define vec(t) vec_new(sizeof(t))

typedef struct {
  void *data;
  size_t el_size;
  size_t capacity;
  size_t size;
} vector;

vector vec_new(size_t el_size);                             
vector vec_with_capacity(size_t el_size, size_t capacity);
int vec_reserve(vector *v, size_t new_capacity);            
int vec_push(vector *v, const void *el);                    
int vec_pop(vector *v, void *el);
int vec_get(const vector *v, size_t i, void *el);
void vec_clear(vector *v);
void vec_delete(vector *v);
```

La implementación de las funciones no es muy relevante, todo funciona correctamente. El tema viene al usarlo, sobre todo para tipos primitivos es un tanto incómodo:

```c
vector nums = vec(int);

int temp = 1;
vec_push(&nums, &temp); // nums contiene 1

temp = 2;
vec_push(&nums, &temp); // nums contiene 1, 2

temp = 3;
vec_push(&nums, &temp); // nums contiene 1, 2, 3

int el;
vec_get(&nums, 1, &el); // el = 2
```

Lo ideal sería poder agregar valores al vector y leer directamente, sin usar una variable auxiliar por medio:

```c
vector nums = vec(int);

vec_push(&nums, 1);
vec_push(&nums, 2);
vec_push(&nums, 3);

int el = vec_get(&nums, 1);
```

El problema es que sin información de tipo, no se pueden hacer funciones de este estilo, como no se puede dereferenciar `void*` sin castearlo explícitamente hay que copiar los elementos con `memcpy`. Por ahora lo resolví haciendo funciones para distintos tipos primitivos:

```c
int pushi(vector *v, int el) {
  int *p_el = &el;
  return vec_push(v, p_el);
}

int popi(vector *v) {
  int el;
  vec_pop(v, &el);
  return el;
}

int vgeti(const vector *v, size_t i) {
  int el;
  vec_get(v, i, &el);
  return el;
}

// repetidas para unsigned, long, char, float, etc...
```

Mi consulta es: esta solución parece poco elegante, por un lado porque es mucha repetición de código idéntico que sólo cambia en un tipo, por otro lado porque no sirve para otros elementos (por ejemplo un vector de structs). Hay alguna forma de hacer funciones de este estilo, o alguna macro, que funcionen con un tipo genérico?

Tengo una macro para obtener un elemento por índice, pero necesita pasarle el tipo cada vez, que tampoco me parece ideal:

```c
#define el(v, t, i) (((t *)((v).data))[(i)]) // uso: int x = el(vec, int, 4);
```

**Comment by marcelogarberoglio** [comment]
En general no se suele usar un vector genérico para almacenar datos porque como vos te diste cuenta es muy engorroso tener que estar casteando para absolutamente todas las operaciones. Además estás asumiendo que el vector será de algún tipo simple (int, float, etc.) pero no serviría para vectores de vectores o vectores de sctructs (las posibilidad son infinitas)

Lo que sí se puede hacer son funciones genéricas, que permitan trabajar con cualquier tipo de vector (pero el usuario sabe si es int*, double* o un vector de vectores, o de structs, etc.) Esas funciones deben recibir -además del vector- el tamaño en bytes de cada elemento y, en algunos casos, una función que permita comparar o analizar los elementos. Un ejemplo es la función qsort de la biblioteca estándar, una función que ordena vectores de cualquier tipo.
Durante el cuatrimestre no vamos a ver este tema, tal vez en alguna clase extra al final del cuatrimestre. Lo que sí vamos a ver es una alternativa más simple para manejar conjuntos de datos genéricos, que es usar typedef.

**Comment by marcelogarberoglio** [comment]
Exacto, la idea es tener funciones genéricas. También -como vamos a ver más adelante- se pueden crear bibliotecas que le permitan al usuario almacenar datos genéricos (conjuntos de enteros, double, structs, etc.), donde la interface tendrá funciones para acceder, agregar, borrar, etc. Pero la forma en que se manipulan esos datos estará oculta para el usuario.

Me alegro de que te gusten estos temas, si querés podés seguir mandando "(fuera de tema)". Dentro de mis posibilidades y conocimientos intentaré contestarlas.
